File: solutions/go/need-for-speed/1/need_for_speed.go
package speed

import (
    "fmt"
)

type Car struct {
    battery int
    batteryDrain int
    speed int
    distance int
}

func NewCar(speed, batteryDrain int) Car {
    return Car{battery: 100, batteryDrain: batteryDrain, speed: speed}
}

type Track struct {
    distance int
}

func NewTrack(distance int) Track {
    return Track{distance: distance}
    
}

func Drive(car Car) Car {
	if car.battery < car.batteryDrain {
        fmt.Println("Not enought battery to drive")
        return car
    }
    car.distance += car.speed
    car.battery -= car.batteryDrain
    return car
}

func CanFinish(car Car, track Track) bool {
	maxDrives := car.battery / car.batteryDrain
    maxDistance := maxDrives * car.speed
    return maxDistance >= track.distance
}

File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
class RemoteControlCar
{
    private int speed;
    private int battery;
    private int batteryDrain;
    private int distanceDriven;
    
    public RemoteControlCar(int speed, int batteryDrain)
    {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
        this.distanceDriven = 0;
        this.battery = 100;
    }

    public int Speed => speed;
    public int Battery => battery;
    public int BatteryDrain => batteryDrain;
    
    public bool BatteryDrained() => battery < batteryDrain;

    public int DistanceDriven() => distanceDriven;

    public void Drive()
    {
        if (!BatteryDrained() && battery >= batteryDrain)
        {
            distanceDriven += speed;
            battery -= batteryDrain;
        }
        else
        {
            battery = 0;
        }
    }

    public static RemoteControlCar Nitro() => new RemoteControlCar(50, 4);
}

class RaceTrack
{
    private int distance;
    
    public RaceTrack(int distance)
    {
        this.distance = distance;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        int maxDrives = car.Battery / car.BatteryDrain;
        int maxDistance = maxDrives * car.Speed;
        return maxDistance >= distance;
    }
}
